### PR TITLE
[DEE-417] do not calculate english-only text properties for not english samples 

### PR DESCRIPTION
**Pull request by yromanyshyn**

### CODE REVIEW

**Review by noamzbr — COMMENTED** [review]
Can you please add some tests?

File: deepchecks/nlp/utils/text_properties.py [hunk 1]
```diff
@@ -236,6 +246,7 @@ def language(raw_text: Sequence[str],
     predictions = model.predict(list(raw_text), k=1, threshold=lang_certainty_threshold)
 
     # x is empty for detection below threshold
+    # TODO: why nan and not None?
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_properties.py]: replaced with None

File: deepchecks/nlp/utils/text_properties.py [hunk 2]
```diff
@@ -412,38 +429,63 @@ def average_sentence_length(raw_text: Sequence[str]) -> List[str]:
     {'name': 'Average Sentence Length', 'method': average_sentence_length, 'output_type': 'numeric'},
 )
 
-LONG_RUN_PROPERTIES = ['Toxicity', 'Fluency', 'Formality', 'Unique Noun Count']
-ENGLISH_ONLY_PROPERTIES = ['Sentiment', 'Subjectivity', 'Toxicity', 'Fluency', 'Formality']
+
+LONG_RUN_PROPERTIES = ('Toxicity', 'Fluency', 'Formality', 'Unique Noun Count')
+ENGLISH_ONLY_PROPERTIES = ('Sentiment', 'Subjectivity', 'Toxicity', 'Fluency', 'Formality')
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_properties.py]: example

```
# %%
text = [
    'Hello world',
    'Explicit is better than implicit',
    'This is a test sentence.',
    'гарна погода',
    'Львів красиве місто',
    'червоний килим'
]
# %%
p = calculate_default_properties(
    raw_text=text,
    include_properties=[
        'Unique Noun Count',
        'Readability Score',
        'Lexical Density',
        'Average Sentence Length'
    ],
    include_long_calculation_properties=True
)
# %%
p[0]
# {'Lexical Density': [100.0, 100.0, 100.0, 100.0, 100.0, 100.0],
#  'Unique Noun Count': [2, 2, 2, 2, 2, 2],  # incorrect for non-eng samples
#  'Readability Score': [77.905, 83.32, 66.4, nan, nan, nan],
#  'Average Sentence Length': [2.0, 5.0, 5.0, 2.0, 3.0, 2.0]}
```

added `Unique Noun Count` and `Readability Score` to the `ENGLISH_ONLY_PROPERTIES` list

**Comment by yromanyshyn** [comment]
@noamzbr 

> Can you please add some tests?

added

File: makefile [hunk 3]
```diff
@@ -174,7 +174,7 @@ vision-torch-tf-setup: env
 
 nlp-tests-setup: env
 	@echo "####  installing nlp properties packages #### "
-	$(PIP) install -q "langdetect>=1.0.9" "textblob>=0.17.1";
+	$(PIP) install -q "langdetect>=1.0.9" "textblob>=0.17.1" "fasttext>=0.8.0"
```
Comment (noamzbr) [inline makefile:177]: I think I've added these to nlp-properties. Do we need this individual setup or can we use the nlp-properties req file?

File: tests/nlp/checks/train_test_validation/property_drift_test.py [hunk 4]
```diff
@@ -108,9 +107,9 @@ def test_without_drift(self, small_wikiann_train_test_text_data):
         assert_that(result.value, has_entries({
             'Text Length': has_entries({'Drift score': 0.0, 'Method': 'Kolmogorov-Smirnov'}),
             '% Special Characters': has_entries({'Drift score': 0.0, 'Method': 'Kolmogorov-Smirnov'}),
-            'Sentiment': has_entries({'Drift score': 0.0, 'Method': 'Kolmogorov-Smirnov'}),
+            'Sentiment': has_entries({'Drift score': None, 'Method': None}),
```
Comment (noamzbr) [inline tests/nlp/checks/train_test_validation/property_drift_test.py:110]: why has this changed to None? Too many samples that are not in english in wikiann?

File: tests/nlp/checks/train_test_validation/property_drift_test.py [hunk 5]
```diff
@@ -140,8 +139,8 @@ def test_with_drift(self, small_wikiann_train_test_text_data):
             'Average Word Length': has_entries({'Drift score': 0.1, 'Method': 'Kolmogorov-Smirnov'}),
             '% Special Characters': has_entries({'Drift score': 0.16000000000000003, 'Method': 'Kolmogorov-Smirnov'}),
             'Text Length': has_entries({'Drift score': 0.30000000000000004, 'Method': 'Kolmogorov-Smirnov'}),
-            'Subjectivity': has_entries({'Drift score': 0.14, 'Method': 'Kolmogorov-Smirnov'}),
-            'Sentiment': has_entries({'Drift score': 0.08000000000000007, 'Method': 'Kolmogorov-Smirnov'})
+            'Subjectivity': has_entries({'Drift score': None, 'Method': None}),
```
Comment (noamzbr) [inline tests/nlp/checks/train_test_validation/property_drift_test.py:142]: same question here

File: deepchecks/nlp/utils/text_properties.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/utils/text_properties.py]: great!

File: tests/nlp/checks/train_test_validation/property_drift_test.py (hunk 4 above)
Comment (yromanyshyn) [reply tests/nlp/checks/train_test_validation/property_drift_test.py:110]: > Too many samples that are not in english in wikiann?

yeap, many samples were not identified as english, and as a result `not enough samples` for drift calculation

File: tests/nlp/checks/train_test_validation/property_drift_test.py (hunk 5 above)
Comment (yromanyshyn) [reply tests/nlp/checks/train_test_validation/property_drift_test.py:142]: same answer)

File: makefile (hunk 3 above)
Comment (yromanyshyn) [reply makefile:177]: we can, but the `nlp-prop-requirements.txt` file does not contain this one - `langdetect>=1.0.9`,
should I add it, or it is unneeded? As I see it is not used anymore?

Comment (noamzbr) [reply makefile:177]: We don't need langdetect anymore, just fasttext

File: tests/nlp/checks/train_test_validation/property_drift_test.py (hunk 4 above)
Comment (noamzbr) [reply tests/nlp/checks/train_test_validation/property_drift_test.py:110]: mm note to self to try to improve that...

**Review by noamzbr — APPROVED** [review]

**Review by Nadav-Barak — COMMENTED** [review]
I wasnt able to verify, but important that we return np.nan for each non english text for the relevant properties (and not skip them)

File: makefile (hunk 3 above)
Comment (yromanyshyn) [reply makefile]: fixed

File: deepchecks/nlp/utils/text_properties.py [hunk 6]
```diff
@@ -201,29 +202,38 @@ def percentage_special_characters(raw_text: Sequence[str]) -> List[float]:
 
 def max_word_length(raw_text: Sequence[str]) -> List[int]:
     """Return list of integers of max word length."""
-    return [max([len(word) for word in text.split()]) for text in raw_text]
+    result = []
+    for text in raw_text:
+        words = text.split()
+        if not words:
+            continue
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_properties.py]: yeap, thanks! fixed

File: deepchecks/nlp/utils/text_properties.py [hunk 7]
```diff
@@ -482,47 +535,90 @@ def calculate_default_properties(
...
+
+        text_properties = new_text_properties
+
+    warning_message = (
+        'Failed to calculate property {0}. '
+        'Dependencies required by property are not installed. '
+        'Error:\n{1}'
+    )
+    # TODO: refactor
+    for prop in text_properties:
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/nlp/utils/text_properties.py]: what exactly? comment?

[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/nlp/utils/text_properties.py]: Just cause i saw 2 for loops of for prop in text_properties + TODO wanted to make sure its not a missed duplicate

File: makefile [hunk 8]
```diff
@@ -227,13 +233,17 @@ test: requirements dev-requirements
...
+
+
 test-win:
 	@test -d $(WIN_ENV) || python -m venv $(WIN_ENV)
 	@$(WIN_ENV)\Scripts\activate.bat
 	$(PIP_WIN) install -q\
 			"torch==1.10.2+cpu" "torchvision==0.11.3+cpu" \
 			-f https://s3.amazonaws.com/pytorch/whl/torch_stable.html;
-	@$(PIP_WIN) install -q "langdetect>=1.0.9" "textblob>=0.17.1";
+	@$(PIP_WIN) install -q "langdetect>=1.0.9" "textblob>=0.17.1" "fasttext>=0.8.0";
```
Comment (noamzbr) [inline makefile:246]: remove langdetect.
Also, maybe we can just add textblob to nlp-prop-requirements and then do install -r on that file here?

File: deepchecks/nlp/utils/text_properties.py [hunk 9]
```diff
@@ -233,10 +244,17 @@ def language(raw_text: Sequence[str],
         raise exp
 
     # Predictions are the first prediction (k=1), only if the probability is above the threshold
-    predictions = model.predict(list(raw_text), k=1, threshold=lang_certainty_threshold)
-
-    # x is empty for detection below threshold
-    language_codes = [x[0].replace('__label__', '') if x else np.nan for x in predictions[0]]
+    predictions = [
+        model.predict(it, k=1, threshold=lang_certainty_threshold)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:248]:
```suggestion
        model.predict(it.replace('\n', ' '), k=1, threshold=lang_certainty_threshold)
```

Comment (Nadav-Barak) [reply deepchecks/nlp/utils/text_properties.py:248]: Otherwise it fails

File: makefile (hunk 8 above)
Comment (yromanyshyn) [reply makefile]: removed this line completely,

>Also, maybe we can just add textblob to nlp-prop-requirements and then do install -r on that file here?

it is already in `nlp-requirements.txt`

**Review by Nadav-Barak — APPROVED** [review]

**Pull request merged**